src/main/java/org/elasticsearch/search/highlight/HighlightPhase.java

```diff
@@ -116,9 +119,26 @@ public void hitExecute(SearchContext context, HitContext hitContext) {
 
                 Query highlightQuery = field.fieldOptions().highlightQuery() == null ? context.parsedQuery().query() : field.fieldOptions().highlightQuery();
                 HighlighterContext highlighterContext = new HighlighterContext(fieldName, field, fieldMapper, context, hitContext, highlightQuery);
-                HighlightField highlightField = highlighter.highlight(highlighterContext);
-                if (highlightField != null) {
-                    highlightFields.put(highlightField.name(), highlightField);
+
+                if (highlighter.cannotHighlight(fieldMapper) && fieldNameContainsWildcards) {
```

Comment:
I see, I was wonder if it even makes sense to call highlight when we know that the source is not available, but I'm forgetting custom highlighters that might want to retrieve field values from other places, so this should still be implementation dependent, I think it should stay as-is.

I think we should already have dedicated tests for fields that have no source and are not stored, can you double check?